### PR TITLE
[CP] Pluggable CP attention via cp_method + attention_sharding

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #3978
* #3977

**Summary**
There are various CP implementations: allgather, ring attention, selective gather, ulysses, and more. The first three are sequence-sharded with different implementation optimizations, while ulysses is a sequence-sharded to head-sharded CP.

To properly support different implementations, we implement the replacement logic within TorchTitan's Config system -- based on the cp_method argument, the logic replaces the inner attention sharding config or even replaces the entire inner attention config. Examples of the former are the allgather-based CP (which we already have for FlexAttention) and the Ulysses CP introduced in this PR. Examples of the latter are selective gather and ring attention, which will come later.

This PR implements this logic and introduces Ulysses CP as an example to show how the replacement works.

**GPT-OSS + Ulysses CP**
Ulysses CP changes the Q, K, V sharding from S(1)@ CP to S(2)@ CP. However, GPT-OSS's attention sink is R@ CP, so this will result in a mismatch (apply_sink is in the local region so no resharding will happen). One quick (but wrong) solution is to change the attention sink to S(2)@ CP as well. This is wrong because FSDP expects all parameters to be R@DP and R@ CP. The spmd_types type check will error out.

We currently raise an NotImplementedError if Ulysses CP is used with GPT-OSS. Fixing this issue will be in a follow-up PR.

### CODE REVIEW

File: torchtitan/models/llama3/parallelize.py [hunk 1]
```diff
@@ -40,6 +41,8 @@ def parallelize_llama(
     NOTE: The passed-in model preferably should be on meta device. Otherwise,
     the model must fit on GPU or CPU memory.
     """
+    validate_context_parallel(model, parallel_dims, parallelism)
```
Comment (tianyu-l) [inline torchtitan/models/llama3/parallelize.py:44]: this can be done in Trainer.Config post_init at config time

File: torchtitan/config/configs.py [hunk 2]
```diff
@@ -208,12 +208,21 @@ class ParallelismConfig:
     context_parallel_degree: int = 1
     """Context parallelism degree. 1 means disabled."""
 
+    context_parallel_method: Literal["allgather", "ulysses"] = "allgather"
+    """
+    Context-parallel attention method:
+    - "allgather": sequence-shard Q, all-gather K/V (default behavior).
+    - "ulysses": head-shard via all-to-all around a standard attention kernel.
+    """
```
Comment (tianyu-l) [inline torchtitan/config/configs.py:216]: Right now it seems pure sharding config options, just like enable_sp or not on TP axis.

I'm a bit worried when we introduce other variants which config couldn't handle, thus requiring `inner_attention` replacement. In those cases, how should we do config? We talked about avoiding introducing strings in config, as it creates coupling -- every time we introduce a new feature, we need to update the options, and it won't be possible for people who extend torchtitan out-of-repo.

Alternatively, I think we can do a `ModelConfigConverter` which is essentially a "config pass", and swaps existing inner attention with CP-version of inner attention.

If that's the case, maybe we should have a default (e.g. all-gather) and let different CP be different ConfigConverters. Ulysses and any out-of-repo CP style can be achieved in consistent ways.

One could argue SP can also be achieved that way. I don't know, maybe SP vs. TP is canonical enough, but clearly people have more than two ways of doing CP.

Comment (fegin) [reply torchtitan/config/configs.py:216]: > I'm a bit worried when we introduce other variants which config couldn't handle, thus requiring inner_attention replacement.

As discussed in the summary, this is designed to achieve this goal. We will simply replace the inner_attention config. Right now  the two methods just replace the inner_attention sharding config. But in the case you mention, we will 1) implement an inner attention module, e.g., SelectiveGatherInnerAttention 2) replace the inner attention with it.

> and it won't be possible for people who extend torchtitan out-of-repo.


CP is a bit different from SP/TP because it is also coupled with dataloader (or trainer). So it is not easy to achieve out-of-repo modification for CP. So model config pass may not be enough for CP. 

And isn't this the reason why you introduce override? Override can achieve the same thing.

> Alternatively, I think we can do a ModelConfigConverter which is essentially a "config pass", and swaps existing inner attention with CP-version of inner attention.

What's the difference between `ModelConfigConverter` and `override`? Also, the core logic of this PR is just doing the config replacement. It's just that Ulysses only need replacing the sharding config, not the entire inner attention config.

I'm worried that there are too many ways for users and TorchTitan developers to replace config. We already used some config programming to replace the MoE dispatcher, but the replacement logic is purely config programming. This PR also does config programming but exposes the nob to CLI. `override` is one step more to consume external modules and perform pattern match replacement.

I actually think the best way (from my perspective) for users to do out-of-repo modification is to create their own model registries. They can do the truly config programming to replace any module configs they want.

File: torchtitan/models/llama3/parallelize.py (hunk 1 above)
Comment (jinsooihm) [reply torchtitan/models/llama3/parallelize.py:44]: Wouldn't it be better to do it in the model's `update_from_config`? There is already a similar check here: https://github.com/pytorch/torchtitan/blob/4c6481182c1c3d7815e3a119a271eebead94c71a/torchtitan/models/common/decoder.py#L158-L172

Comment (jinsooihm) [reply torchtitan/models/llama3/parallelize.py:44]: Maybe a noob question, but why are there many models (llama3, deepseek v3, gpt_oss, qwen3) that call `validate_config` function in `parallelize.py`? It seems like that can also be moved to `update_from_config`.

Comment (tianyu-l) [reply torchtitan/models/llama3/parallelize.py:44]: we should, also I'm looking for a way to consolidate `update_from_config` and `Trainer.Config.__post_init__`

File: torchtitan/config/configs.py (hunk 2 above)
Comment (tianyu-l) [reply torchtitan/config/configs.py:216]: > And isn't this the reason why you introduce override? Override can achieve the same thing.

> What's the difference between ModelConfigConverter and override?

Latest thoughts on override:
One should only use `override` for two things
1. you are not torchtitan team, but you want to pip install titan, and patch your modules and maybe publish a compatible library so that people can jointly pip install. E.g. you are a hardware vendor, with every titan release, you release a compatible package, making it fast on your hardware.
2. you are torchtitan team, you are too tired of "people complaining about titan perf who're unwilling to do any kernel work", but as titan team member you have high bar on code quality / organization -- you are willing to patch some thing in to showcase perf but you don't want to maintain the code. Current `torchtitan/overrides/` belong to this category.

> We already used some config programming to replace the MoE dispatcher, but the replacement logic is purely config programming.

The current `ModelConfigConverter` is just a config space helper. I'm not opinionated about how people construct the eventual config (in `model_registry` or `config_registry`), but as a maintainer I have to have some way to constructing config and the current ways are convenient for me.

> I actually think the best way (from my perspective) for users to do out-of-repo modification is to create their own model registries. They can do the truly config programming to replace any module configs they want.

Yes I agree. If you don't need to maintain torchtitan to support multiple models + multiple quantization + lora + CP, you can do it however you want.

> So model config pass may not be enough for CP.

Like I said, it's just convenient for current use case. For CP with dataloader modifcation, we can have a `TrainerConfigConverter` -- it's just config programming util.

> I'm worried that there are too many ways for users and TorchTitan developers to replace config.

As mentioned above, maybe all via `*ConfigConverter` as long as we agree on something. This is also the reason why I hesitate to agree with what this PR is doing, a.k.a.

> Also, the core logic of this PR is just doing the config replacement. It's just that Ulysses only need replacing the sharding config, not the entire inner attention config.

I think for Ulysses it seems fine (given how similar it is to TP/SP). Could you share what's the plan for others, e.g. "ring"? My guess is that it'll be similar to quantization, for which we are doing `ModelConfigConverter` today but we could also have args in `model_registry(..., linear_quantization_type, moe_quantization_type, ...)`.

Also I'm not yet convinced that it's helpful to expose `context_parallel_methods \in ["allgather", "ulysses", "selective_gather", "ring"]` to anyone who knows `ParallelismConfig`. This sounds information leakage.

> This PR also does config programming but exposes the nob to CLI.

I think this is not that important
- our quantization doesn't support CLI as well
- if one really want CLI capability, they should build on top of torchtitan configs and expose higher level CLI options

File: torchtitan/models/common/attention_sharding.py [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    """
+
+    seq = SpmdLayout(
+        {DP: spmd.V, CP: spmd.V, TP: spmd.V},
+        partition_spec=(DP, CP, TP, None),
+    )
+    head = SpmdLayout(
+        {DP: spmd.V, CP: spmd.V, TP: spmd.V},
+        partition_spec=(DP, None, (TP, CP), None),
+    )
```
Comment (pianpwk) [inline torchtitan/models/common/attention_sharding.py:70]: it's cool this is expressible

File: torchtitan/models/common/attention_sharding.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    n_kv_heads = getattr(attention_cfg, "n_kv_heads", None)
+    if n_kv_heads is None:
+        n_kv_heads = n_heads
+    if n_heads % div or n_kv_heads % div:
+        raise ValueError(
+            "Ulysses context parallel shards the head dim on the TP and CP axes, "
+            f"so n_heads ({n_heads}) and n_kv_heads ({n_kv_heads}) must both be "
+            f"divisible by tensor_parallel_degree * context_parallel_degree "
+            f"({tp} * {cp} = {div})."
+        )
```
Comment (pianpwk) [inline torchtitan/models/common/attention_sharding.py:141]: sorry, maybe dumb q, could you remind me why this doesn't work

**Comment by fegin** [comment]
Close this in favor of https://github.com/pytorch/torchtitan/pull/4322.
